lib/ex_unit/lib/ex_unit/pattern_diff.ex

```diff
@@ -0,0 +1,421 @@
+    {ast_result, env} = compare(ast, rh_value, env)
+    when_result = evaluate_when(clause, env)
+    {%ContainerDiff{type: :when, items: [ast_result, when_result]}, env}
+  end
+
+  def compare(%{ast: {:=, _, [l, r]}}, rh_value, env) do
+    # Assignment in a pattern, identify which is the binding and which is the source data
+    val =
+      case l do
+        {_, _, nil} -> r
+        _ -> l
+      end
+
+    compare(%{ast: val}, rh_value, env)
+  end
+
+  def compare(%{ast: {:_, _, _}} = pattern, rh_value, env) do
+    {
+      %__MODULE__{
+        type: :value,
+        lh: pattern,
+        rh: rh_value,
+        diff_result: :eq
+      },
+      env
+    }
+  end
+
+  def compare(%{ast: {var_name, _, var_ctx} = var} = pattern, rh_value, {vars, pins} = env)
+      when is_atom(var_name) and is_atom(var_ctx) do
+    var = get_var(var)
+
+    case Map.get(vars, var, :ex_unit_unbound_var) do
+      :ex_unit_unbound_var ->
+        new_vars = Map.put(vars, var, rh_value)
+
+        {
+          %__MODULE__{
+            type: :value,
+            lh: pattern,
+            rh: rh_value,
+            diff_result: :eq
+          },
+          {new_vars, pins}
+        }
+
+      ^rh_value ->
+        {
+          %__MODULE__{
+            type: :value,
+            lh: pattern,
+            rh: rh_value,
+            diff_result: :eq
+          },
+          env
+        }
+
+      _other_value ->
+        {
+          %__MODULE__{
+            type: :value,
+            lh: pattern,
+            rh: rh_value,
+            diff_result: :neq
+          },
+          env
+        }
+    end
+  end
+
+  def compare(%{ast: lh_value} = pattern, rh_value, env)
+      when is_integer(lh_value) and is_integer(rh_value) do
+    like_value_compare(pattern, rh_value, env)
+  end
+
+  def compare(%{ast: lh_value} = pattern, rh_value, env)
+      when is_float(lh_value) and is_float(rh_value) do
+    like_value_compare(pattern, rh_value, env)
+  end
+
+  def compare(%{ast: lh_value} = pattern, rh_value, env)
+      when is_binary(lh_value) and is_binary(rh_value) do
+    like_value_compare(pattern, rh_value, env)
+  end
+
+  def compare(%{ast: lh_value} = pattern, rh_value, env)
+      when is_atom(lh_value) and is_atom(rh_value) do
+    like_value_compare(pattern, rh_value, env)
+  end
+
+  def compare(%{ast: _lh_value} = pattern, rh_value, env) do
+    {
+      %__MODULE__{
+        type: :different,
+        lh: pattern,
+        rh: rh_value,
+        diff_result: :neq
+      },
+      env
+    }
+  end
+
+  def compare(@no_value, rh_value, env) do
+    {
+      %__MODULE__{
+        type: :different,
+        lh: @no_value,
+        rh: rh_value,
+        diff_result: :neq
+      },
+      env
+    }
+  end
+
+  defp compare_list(%{ast: [{:|, _, [head, tail]} | _rest]}, [rh_head | rh_tail], env) do
+    head_pattern = %{ast: head, type: :cons_l}
+    tail_pattern = %{ast: tail, type: :cons_r}
+    {h, env} = compare(head_pattern, rh_head, env)
```

Comment:
@josevalim I don't know if it's the case, but I think it would be amazing if the `List.myers_difference` could also accept a comparison function since it's now accepting a diff one.